### PR TITLE
core: reset tx lookup cache if necessary

### PR DESCRIPTION
**Pull request by rjl493456442**
This pull request resets the txlookup cache if certain transactions are reorg'd out, preventing them from remaining reachable. It addresses failures in the hive tests.

### CODE REVIEW

**Comment by lightclient** [comment]
So it looks like before #28703, the eth api backend would call `rawdb.ReadTransaction(..)` to retrieve txs to serve `eth_getTransactionReceipt`. That function would then check for a `txLookupEntry` (which was deleted in the code that this PR is touching) and if not found, return `nil`.

In #28703, the eth api backend was modified to immediately call `GetTransactionLookup(..)` which first checks the lookup cache before it falls back to the db.

That is why purging the cache upon reorgs resolves the hive test issues.

--

It seems like there are three was to proceed:
1) merge this PR and purge the cache on reorgs
2) revert the logic change to the eth api backend and continue using `rawdb.ReadTransaction(..)`
3) modify the hive test to get the tx receipt, then query the number to see if the canonical hash matches the block has the tx was mined in

I think my preference is 1) or 2). This is a breaking change in terms of API behavior so unless there is reason to retain the new behavior I think we should return to the old behavior.

**Comment by lightclient** [comment]
Also the test this PR is failing seems to be flaky.

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM, the txlookupcache is only used by the api, so the cost of purging the cache only really impacts nodes that do a lot of transaction lookups (due to tracing or rpc). Normal nodes should not be impacted by this change at all

**Comment by vr-devil** [comment]
waiting a release online ~ come on ~

File: core/blockchain.go [hunk 1]
```diff
@@ -2202,11 +2208,13 @@ func (bc *BlockChain) reorg(oldHead *types.Header, newHead *types.Block) error {
 
 	// Delete useless indexes right now which includes the non-canonical
 	// transaction indexes, canonical chain indexes which above the head.
-	indexesBatch := bc.db.NewBatch()
-	for _, tx := range types.HashDifference(deletedTxs, addedTxs) {
+	var (
+		indexesBatch = bc.db.NewBatch()
+		diffs        = types.HashDifference(deletedTxs, addedTxs)
+	)
+	for _, tx := range diffs {
```
Comment (lightclient) [inline core/blockchain.go:2215]: If we're going to purge before writing canonical chain, can we revert this chunk?

Comment (rjl493456442) [reply core/blockchain.go:2215]: They are equal, i am just too lazy to wait another round of CI. I think it should be fine.

**Pull request merged**